std::alloc: use posix_memalign instead of memalign on solarish

devnexen opened the pull request:
`memalign` on Solarish requires the alignment to be at least the size of a pointer, which we did not honor. `posix_memalign` also requires that, but that code path already takes care of this requirement.

close GH-124787

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-17`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/8969965559/job/24632460192) [(plain)](https://github.com/rust-lang/rust/commit/3fddb3bcb375c630e004dec20726ef743503e1ee/checks/24632460192/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#16 exporting to docker image format
#16 sending tarball 29.1s done
#16 DONE 33.8s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-17]
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-17', '--enable-llvm-link-shared', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'change-id=99999999', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-17/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.thin-lto-import-instr-limit := 10
configure: change-id            := 99999999
```

</details>

devnexen committed:
std::alloc: using posix_memalign instead of memalign on solarish.

simpler code path since small alignments are already taking care of.
close GH-124787

klensy commented on library/std/src/sys/pal/unix/alloc.rs:
```diff
@@ -87,7 +87,11 @@ cfg_if::cfg_if! {
+            //
+            // We purpose use the higher alignment so we have it at least
+            // a proper word size value guaranteed.
+            let align = layout.align().max(create::mem::size_of::<usize>());
```
```suggestion
            let align = layout.align().max(crate::mem::size_of::<usize>());
```

RalfJung commented:
Can't we just use `posix_memalign`? [Solaris](https://docs.oracle.com/cd/E88353_01/html/E37843/posix-memalign-3c.html) and [Illumos](https://illumos.org/man/3C/posix_memalign) seem to both have that. Then the diff would be just two lines, as we're already using posix_memalign by default when we don't have some other special case.

devnexen commented:
If we do not care of solaris < 10 then it s possible indeed.

RalfJung commented:
https://doc.rust-lang.org/rustc/platform-support.html explicitly says "Solaris 11" so I think indeed we do not care about Solaris 10.

RalfJung commented:
LGTM, thanks!
@bors r+

bors commented:
:pushpin: Commit 1e232fea1b0e5bcf43230a8a28a24f9e2cbac988 has been approved by `RalfJung`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=RalfJung 1e232fea1b0e5bcf43230a8a28a24f9e2cbac988 -->
<!-- homu: {"type":"Approved","sha":"1e232fea1b0e5bcf43230a8a28a24f9e2cbac988","approver":"RalfJung","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 1e232fea1b0e5bcf43230a8a28a24f9e2cbac988 with merge b71fa82d786ae1b5866510f1b3a7e5b7e1890e4c...
<!-- homu: {"type":"BuildStarted","head_sha":"1e232fea1b0e5bcf43230a8a28a24f9e2cbac988","merge_sha":"b71fa82d786ae1b5866510f1b3a7e5b7e1890e4c"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/9051010363/job/24868697375)
Approved by: RalfJung
Pushing b71fa82d786ae1b5866510f1b3a7e5b7e1890e4c to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"RalfJung","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/9051010363/job/24868697375"},"merge_sha":"b71fa82d786ae1b5866510f1b3a7e5b7e1890e4c"} -->
